### ISSUE TITLE
Math ops support for float64

### ISSUE THREAD

**Issue by smcantab** [tensorflow/tensorflow#1140]
I am raising this issue following [this stackoverflow discussion](https://stackoverflow.com/questions/35443080/tensorflow-critical-graph-operations-assigned-to-cpu-rather-than-gpu?lq=1). I found that there is not that much support for float64 ops, that is needed for some numerical applications. A number of other issues  #1061 #761 #547 and PR #1089 have been raised around this.

On the basis of how my graph is initialised, things like **MatMul** (also raised in this other [stackoverflow question](https://stackoverflow.com/questions/35428297/tensorflow-issue-with-gpu-on-matmul-gpu-isnt-recognized?lq=1)), **Mean, L2Loss, Mean_grad/Prod, Mul_grad/Sum, AddN, SoftmaxCrossEntropyWithLogits** are not implemented for float64.

Would it be a good idea to extend PR #1089 to more ops?

**Comment by girving** [comment]
Looks like #1089 was already merged, so we'll need more PRs.

**Comment by siddharth-agrawal** [comment]
I want to start contributing to Tensorflow, so thought this issue would be a good start. I started with MatMul from among the ops mentioned by @smcantab. In the op implementation I found [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/matmul_op.cc#L210), float64 seems to be purposely commented out. Is there a reason for this?

**Comment by smcantab** [comment]
as @dave-andersen pointed out [here](https://stackoverflow.com/questions/35443080/tensorflow-critical-graph-operations-assigned-to-cpu-rather-than-gpu) some operations have not been registered for double simply to reduce the bloat in the binaries, given that, as far as I understand, it was not clear whether there'd be much interest in supporting double precision. Obviously there's been quite some interest so we should go ahead. For some operations it should be as simple as registering the kernel for double, and checking that the tests are there. Some other operations might require tailored optimization. I have been busy with some other stuff, so I have not been able to work on a PR yet, but @siddharth-agrawal if you start one I'll follow it and contribute as soon as possible.

**Comment by siddharth-agrawal** [comment]
Sounds good. I will try to get something out by the end of this week.

**Comment by girving** [comment]
@smcantab: Many people have requested better double support and it's critical for a lot of numerical applications, so we should do it even it increases binary size.  complex128 is rarer but the same principle applies.

**Comment by siddharth-agrawal** [comment]
@girving Some of the ops mentioned by @smcantab like **Mean**, **Prod** have float64 support on the GPU as of today. Are there some important ops that need to be looked at?

**Comment by deepmath** [comment]
Here are the ops that lack double versions.  Most of these would be straightforward to add.  The most important missing ones are the max pool ops covered in issue #547, but as described in #547 they'd be a bit difficult to do.

    AdjustContrastv2
    AllCandidateSampler
    AudioSummary
    CTCBeamSearchDecoder
    CTCGreedyDecoder
    CTCLoss
    ComputeAccidentalHits
    DrawBoundingBoxes
    EditDistance
    ExtractGlimpse
    FixedUnigramCandidateSampler
    HSVToRGB
    ImageSummary
    InTopK
    LRN
    LRNGrad
    LearnedUnigramCandidateSampler
    LogUniformCandidateSampler
    MaxPool
    MaxPoolGrad
    MaxPoolGradWithArgmax
    MaxPoolWithArgmax
    NegTrain
    RGBToHSV
    SampleDistortedBoundingBox
    SparseMatMul
    StringToNumber
    TensorArrayConcat
    TensorArrayGrad
    TensorArrayPack
    TensorArrayRead
    TensorArraySize
    TensorArraySplit
    TensorArrayUnpack
    TensorArrayWrite
    ThreadUnsafeUnigramCandidateSampler
    UniformCandidateSampler

**Comment by girving** [comment]
Here are the ops missing double support.  I believe most of these would be straightforward to add, except for max pooling as described in #547, and we'd be happy to accept PRs.  Also max pooling is doable, it would just take a bit of coordination to make the stream executor changes.
 
    AdjustContrastv2
    AllCandidateSampler
    AudioSummary
    CTCBeamSearchDecoder
    CTCGreedyDecoder
    CTCLoss
    ComputeAccidentalHits
    DrawBoundingBoxes
    EditDistance
    ExtractGlimpse
    FixedUnigramCandidateSampler
    HSVToRGB
    ImageSummary
    InTopK
    LRN
    LRNGrad
    LearnedUnigramCandidateSampler
    LogUniformCandidateSampler
    MaxPool
    MaxPoolGrad
    MaxPoolGradWithArgmax
    MaxPoolWithArgmax
    NegTrain
    RGBToHSV
    SampleDistortedBoundingBox
    SparseMatMul
    StringToNumber
    TensorArrayConcat
    TensorArrayGrad
    TensorArrayPack
    TensorArrayRead
    TensorArraySize
    TensorArraySplit
    TensorArrayUnpack
    TensorArrayWrite
    ThreadUnsafeUnigramCandidateSampler
    UniformCandidateSampler

**Comment by siddharth-agrawal** [comment]
@girving Just out of curiosity, is there a way these can be found easily, i.e., not manually?

**Comment by girving** [comment]
@siddharth-agrawal Here's the script I used, but it would have to be modified a bit to work outside of Google: https://gist.github.com/girving/fbe861e6df1ce9d5add8e57bf32a247d

**Comment by yaroslavvb** [comment]
BTW, you can also get op_def which has type-attrs for an op like this `tf.add.__globals__["_op_def_lib"]._ops["Add"].op_def`
There is one _op_def_lib per each gen.*ops.py file, so 19 "_op_def_lib" objects. If only we had some easy method like this to find which ops have GPU kernels....

**Comment by siddharth-agrawal** [comment]
@girving The script fails to catch some of the ops, for example `TensorArrayConcat`, which is registered for float64 [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/tensor_array_ops.cc#L668) and [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/tensor_array_ops.cc#L686). Tests have also been written for it as can be seen [here](https://github.com/tensorflow/tensorflow/commit/01806e89e5dd141ae456bfee2ca0e799fb1aa32d#diff-e95198e570ac301ead1b1640fce0f656). The same holds true for the following:
`TensorArrayPack
TensorArrayRead
TensorArraySplit
TensorArrayUnpack
TensorArrayWrite`

**Comment by siddharth-agrawal** [comment]
@girving `AdjustContrastv2` is used internally, and the op doesn't need more type registrations as far as I can tell.

**Comment by girving** [comment]
@siddharth-agrawal Ah, right.  Deprecated ops definitely shouldn't be extended. :)

**Comment by siddharth-agrawal** [comment]
@girving I think the deprecated op is `AdjustContrast`, not `AdjustContrastv2`. `AdjustContrastv2` is used internally for `adjust_contrast()`, but in a way that takes care of all the data types.

**Comment by ibab** [comment]
I'd like to help with this, as I have a few applications that could benefit from better `float64` support.
@siddharth-agrawal: Are you okay with splitting the work on this to speed things up? We could use this issue to announce which ops we want to work on, so we don't end up duplicating work.

I've noticed that the reduction ops for `float64` are quite slow when executing on the GPU.
I've submitted some changes to Eigen that enable the faster optimized code path:
https://bitbucket.org/eigen/eigen/pull-requests/207/enable-efficient-tensor-reduction-for/diff

**Comment by siddharth-agrawal** [comment]
@ibab Surely, how about I take all the `*Sampler` ones (unless you want `*Sampler`s for your application)? @samjabrahams is already working on `CTC*`, so that leaves quite a few ops.

@samjabrahams I saw the PR you referenced above, it looks like the right change for CPU. However, the issue is to enable `float64` for the GPU, and as far as I remember the `CTC*` ops don't have any GPU implementations. So my guess is that you will have to write GPU implementations for those ops as well. Same is true for the `*Sampler` ops that I'm taking up.

**Comment by samjabrahams** [comment]
Cool, I'll give it a shot.

**Comment by siddharth-agrawal** [comment]
@samjabrahams I should have been more clear above. The issue title suggests that we just need to enable `float64` for the ops. However, if you look at the StackOverflow links in the issue description, you will see what I meant.

**Comment by samjabrahams** [comment]
Definitely- no worries on your end, I should have been more diligent reading the original message.

Something I've noticed is that some Ops have fleshed out GPU implementations, while others simply register to GPU using the same implementation as CPU. Is that due to some operations being able to take better advantage of highly-parallel computation, or am I missing something?

**Comment by siddharth-agrawal** [comment]
@samjabrahams I have the same question, since I haven't written any op implementation. @girving could you give us a brief explanation of the same?

**Comment by ibab** [comment]
@samjabrahams @siddharth-agrawal I think the reason is that if there is no registered GPU kernel for an op, then that op will give you an error if you execute it inside a `with tf.device('/gpu:0'):` block, which is quite annoying.

There are a few situations where it doesn't really make sense to have a real GPU op, and there might be ops for which a proper GPU kernel hasn't been written yet and we want to substitute an existing CPU op.
The solution to the above problem is to register a "fake" GPU kernel that actually just executes the regular CPU op.
This is done by placing all input/output tensors into host memory using the `.HostMemory("x")` kernel registration method and registering the CPU op as the implementation.
We have to use `.HostMemory`, because otherwise TensorFlow will automatically copy the data into GPU memory, where our CPU kernel can't reach it.

A good example that I've found is the `InvertPermutation` kernel here: https://github.com/tensorflow/tensorflow/blob/d42facc3cc9611f0c9722c81551a7404a0bd3f6b/tensorflow/core/kernels/transpose_op.cc#L77
It's expected that it will be called on an extremely small array (with length equal to the rank of some tensor), so we don't want it to be a real GPU kernel, as we want to avoid needlessly copying our tiny array to the device.
At the same time, we want to be able to run inside a `tf.device('/gpu:0')` block, so we make a fake GPU kernel registration.

**Comment by yaroslavvb** [comment]
@ibab another example is integer [Add](https://github.com/tensorflow/tensorflow/blob/d42facc3cc9611f0c9722c81551a7404a0bd3f6b/tensorflow/core/kernels/cwise_op_add.cc#L28) . At some point there were shape-related calculations on Tensors on GPU had a bottleneck in crossing the logical device boundary, so those ops were registered with "fake" GPU implementations to avoid the gpu/cpu device crossing overhead

**Comment by siddharth-agrawal** [comment]
@ibab Thanks for the explanation. However, there are cases where the implementation of the op has the device as a template parameter, without the `HostMemory` bit. Some examples are `[MatMul](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/matmul_op.h)`, `[L2Loss](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/l2loss_op.h)`. I guess the question was: how do we decide if we need to write an explicit GPU implementation? I guess this has something to do with the underlying Eigen interfaces.

**Comment by ibab** [comment]
You will need to instantiate an explicit GPU implementation if you want to actually use the GPU.
`MatMul` for example actually runs on the GPU, so it wouldn't be correct to use `.HostMemory`.

It might be a bit tricky to spot how running GPU computations is accomplished, because it is nicely abstracted by the Eigen interface.
Whenever you see a line like
```cpp
output.device(d) = // ... some expression involving an `input` tensor
```
then this will either run on the host (if `d` is a `CPUDevice`) or will run on a GPU (if `d` is a `GPUDevice`).
You can grab the device by calling `ctx->eigen_device<Device>()`.

For example, the `MatMulOp` calls it's functor with the current device here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/matmul_op.cc#L160
Then this line gets executed:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/matmul_op.h#L38

**Comment by ibab** [comment]
@yaroslavvb: Thanks for clarifying this, I was actually wondering why `int32` was treated as a special case in many places.

**Comment by siddharth-agrawal** [comment]
@ibab Thanks again, that clarifies some things.

**Comment by ebrevdo** [comment]
Pretty sure TensorArray calls all support float64

**Comment by ibab** [comment]
The reason the script flags `TensorArray*` is because of the flow inputs/outputs.
These happen to be `float32` tensors.

**Comment by siddharth-agrawal** [comment]
@girving I did some digging on `LogUniformSampler` to see how it works. I found that the sampling is done using `RandDouble()` which is a part of the class `SimplePhilox`. However, `RangeSampler` which is the general class that `LogUniformSampler` uses, has been written so as to make it work only for the CPU. Is there a reason for this? If not, how should I go about making the changes (some pointers would be good, things that I should keep in mind while making the changes)? Since we have to write GPU implementations, should this be tracked in a different issue (as it is not specifically for `float64`)?

**Comment by girving** [comment]
@siddharth-agrawal If you have to write GPU implementations, that should definitely be a separate issue.

**Comment by siddharth-agrawal** [comment]
@girving Created a separate issue for GPU implementations. Could you comment on the other questions in my comment above?

**Comment by girving** [comment]
@siddharth-agrawal I'm leery of trying to GPUify all the range sampling ops unless you have a compelling need.  They're quite intricate, and specifically intricate in the kind of way that's hard to do efficiently on GPU (doing them inefficiently on GPU is not interesting).  It's all possible, but their has to be a need in order to accept that complexity into the codebase.

**Comment by siddharth-agrawal** [comment]
@girving Alright, could you then comment on which ops should be worked on? This is so that anyone who wants to contribute does not spend time on something that is not useful. For example, @ebrevdo mentioned that float64 for CTC ops may not be useful, and as you said the Sampler ops will also not be supported. Are there any other ops that need to be discarded from [this list](https://github.com/tensorflow/tensorflow/issues/1140#issuecomment-227012645)?

**Comment by girving** [comment]
@siddharth-agrawal I would suggest picking a machine learning project to apply TensorFlow too!  If you pick something that you find interesting, you'll inevitably find flaws or areas for improvement in TensorFlow that will help your project.  Also inevitably, things that will help you will help others, and they'll probably end up more useful than unmotivated changes.

The other options is to look through the contributions welcome list searching for other small projects, but I think that will be less interesting: https://github.com/tensorflow/tensorflow/labels/contributions%20welcome

**Comment by aselle** [comment]
Automatically closing due to lack of recent activity. Please file a new issue referring to this issue, if the issue is still relevant and important. If this is documenting some kind of bug, be sure to use the latest version of TensorFlow. Thank you so much.
